### ISSUE TITLE
Get some Dot Net Core support

### ISSUE THREAD

**Issue by Ornias1993** [Ornias1993/jailman#15]
Currently there is no official Dot Net Core support for FreeBSD.
This limits some of the jails we are able to run.

It would be awesome if someone gets some level of Dot Net Core support into a FreeNAS/FreeBSD/iocage jail...

## References
**Some previous work on this has been done here:**

https://github.com/dotnet/source-build/issues/1139
https://github.com/dotnet/runtime/issues/10355
https://github.com/dotnet/runtime/issues/31039

**Most work has been done by:**

@mateusrodrigues
@wfurt
@jasonpugsley

**Comment by Ornias1993** [comment]
Current status of this:
- There has been unofficial builds of DotNet for FreeBSD for a while now.
- Said builds run on FreeNAS
- Said builds do NOT run in FreeNAS 11.x iocage jails due to requiring mlock access and https://github.com/iocage/iocage/pull/619
- Said builds can get troublesome, because ZFS can exceed the max wirable memory and thus prevent DotNet from using mlock at all. I've reported the bug here: https://jira.ixsystems.com/browse/NAS-106193

**Comment by Ornias1993** [comment]
Blocked due to lack of mlock support within jails on freenas 11.x

**Comment by jasonpugsley** [comment]
I did a quick search of the .NET runtime source and it looks like mlock() is only used for one purpose (though the same code appears in two different files).
[`InitializeFlushProcessWriteBuffers()`](https://github.com/dotnet/runtime/blob/6d1090310818a3b1c6e41dd02f060395cb660391/src/coreclr/src/pal/src/thread/process.cpp#L3520) src/coreclr/src/pal/src/thread/process.cpp
[`GCToOSInterface::Initialize()`](https://github.com/dotnet/runtime/blob/46b430aa0c90698cedd90f71fcad53312c08b732/src/coreclr/src/gc/unix/gcenv.unix.cpp#L307) src/coreclr/src/gc/unix/gcenv.unix.cpp
```c
        // Locking the page ensures that it stays in memory during the two mprotect
        // calls in the FlushProcessWriteBuffers below. If the page was unmapped between
        // those calls, they would not have the expected effect of generating IPI.
        int status = mlock(g_helperPage, OS_PAGE_SIZE);
```
This is for the the use of [`FlushProcessWriteBuffers()`](https://github.com/dotnet/runtime/blob/6d1090310818a3b1c6e41dd02f060395cb660391/src/coreclr/src/pal/src/thread/process.cpp#L3555) and  [`GCToOSInterface::FlushProcessWriteBuffers()`](https://github.com/dotnet/runtime/blob/46b430aa0c90698cedd90f71fcad53312c08b732/src/coreclr/src/gc/unix/gcenv.unix.cpp#L499)

So if someone who understands FreeBSD kernel internals well enough to suggest a different _reliable_ way to generate the `IPI` without the need to lock the page in memory then we could submit a PR to change this function for FreeBSD.

Linux code, for example, is making use of `membarrier(MEMBARRIER_CMD_PRIVATE_EXPEDITED, 0)` instead  of the `mprotect()` solution employed for FreeBSD.

This discussion might help explain the need for `FlushProcessWriteBuffers()` [https://github.com/dotnet/runtime/issues/3952](https://github.com/dotnet/runtime/issues/3952)

**Comment by Ornias1993** [comment]
@jasonpugsley i know, that has been explained to me.But it not relevant for this project. 

As this is just automation scripting, I already forwarded the issue to Allan (Jude), who forwarded it to someone else... 

However: mlock is not a problem anymore with FreeBSD itself, they fixed it already by seperating ZFS wiring from the system limit. Due to an error on the part of IXSystems this was not merged into FreeNAS

In the mean time I got IXSystems to fix the general memory wiring issue on FreeNAS and mlock is added to jails in TrueNAS CORE (v12), which solves the issues I have in a few months anyway.

**Comment by Ornias1993** [comment]
As we are now working on 1.3.0, which will be based on TrueNAS CORE, I consider this "unblocked"
